File: python3/convert_bst_to_greater_tree.py
from .model import TreeNode

# Space : O(1)
# Time  : O(n)


class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:

        def convert(root, total):
            if not root:
                return total

            total = convert(root.right, total)
            root.val += total
            total = convert(root.left, root.val)

            return total

        convert(root, 0)
        return root

File: golang/convert_bst_to_greater_tree.go
package main

// Space	: O(n)
// Time 	: O(n)

import (
	"golang/model"
)

func getInorder(root *model.TreeNode) []int {
	res := []int{}

	if root != nil {
		res = append(res, getInorder(root.Left)...)
		res = append(res, root.Val)
		res = append(res, getInorder(root.Right)...)
	}

	return res
}

func insertGreaterTree(root *model.TreeNode, nodes []int) {
	if root != nil && len(nodes) > 0 {
		insertGreaterTree(root.Left, nodes)
		root.Val = nodes[0]
		nodes = append(nodes[:0], nodes[1:]...)
		insertGreaterTree(root.Right, nodes)
	}
}

func convertBST(root *model.TreeNode) *model.TreeNode {
	nodes := getInorder(root)
	for i := len(nodes) - 2; i >= 0; i-- {
		nodes[i] += nodes[i+1]
	}

	point := root
	insertGreaterTree(point, nodes)
	return root
}
